src/lib/keyseq.ts

```diff
@@ -473,37 +475,44 @@ function numericPrefixToExstrSuffix(numericPrefix: KeyEventLike[]) {
  * translation map must be length-1 strings.
  */
 export function translateKeysUsingKeyTranslateMap(
-    keyEvents: KeyEventLike[],
+    keyEvents: MinimalKey[],
     keytranslatemap: { [inkey: string]: string },
 ) {
     for (let index = 0; index < keyEvents.length; index++) {
         const keyEvent = keyEvents[index]
         const newkey = keytranslatemap[keyEvent.key]
 
-        // KeyboardEvents can't have been translated, MinimalKeys may
-        // have been. We can't add anything to the MinimalKey without
-        // breaking a ton of other stuff, so instead we'll just assume
-        // that the only way we've gotten a MinimalKey is if the key
-        // has already been translated. We err way on the side of
-        // safety becase translating anything more than once would
+        // Translating anything more than once would
         // almost certainly mean oscillations and other super-weird
         // breakage.
-        const neverTranslated = keyEvent instanceof KeyboardEvent
-        if (neverTranslated && newkey !== undefined) {
-            // We can't update the keyEvent in place. However, the
-            // entire pipeline works with MinimalKeys all the way
-            // through, so we just swap the key event out for a new
-            // MinimalKey with the right key and modifiers copied from
-            // the original.
-            keyEvents[index] = new MinimalKey(newkey, {
-                altKey: keyEvent.altKey,
-                ctrlKey: keyEvent.ctrlKey,
-                metaKey: keyEvent.metaKey,
-                shiftKey: keyEvent.shiftKey,
+        if (!keyEvent.translated && newkey !== undefined) {
+            keyEvents[index] = minimalKeyFromKeyboardEventLike({
```

Comment:
minimalKeyFromKeyboardEvent was supposed to convert from keyboardEvemt to MinimalKey, using code if configured, once per event. I don't think it's a good idea to make it general and use here, if we want to be more concise we should either use a separate helper to update or make .key mutable.

Reply:
I don't see the problem. Making key mutable seems dangerous since lots of other things might need to change when the key changes.

This way if we make a new requirement like needing a keycode we'll notice that it is missing from translated keys.

And it just seemed neater.

Reply:
My idea is that we don't need code in internal representation of keys, but we need it to convert to this representation once in the top of the pipeline. We convert either trivially or (after second PR) using code, and then use .key property same as before and apply translations. It doesn't make sense to use conversion function twice, we better create someting like updateKey(), or make .key private and create setter.